tangelo/linq/translator/translate_sympy.py

```diff
@@ -0,0 +1,232 @@
+    from sympy.physics.quantum.gate import UGate
+
+    rz_matrix = ImmutableMatrix([[exp(- I * theta / 2), 0], [0, exp(I * theta / 2)]])
+
+    return UGate(target, rz_matrix)
+
+
+def p_gate(target, theta):
+    """Implementation of the PHASE gate with a unitary matrix.
+
+    Args:
+        target (int): Target qubit.
+        theta (float): Phase parameter.
+
+    Returns:
+        sympy.physics.quantum.gate.UGate: Self-explanatory.
+    """
+
+    from sympy import ImmutableMatrix, I, exp
+    from sympy.physics.quantum.gate import UGate
+
+    p_matrix = ImmutableMatrix([[1, 0], [0, exp(I * theta)]])
+
+    return UGate(target, p_matrix)
+
+
+def controlled_gate(gate_function):
+    """Change a one-qubit gate to a controlled gate.
+
+    Args:
+        gate_function (sympy.physics.quantum.gate): Class for a specific gate.
+
+    Returns:
+        function: Function wrapping a gate with CGate.
+    """
+
+    from sympy.physics.quantum.gate import CGate
+
+    def cgate(control, target, *args, **kwargs):
+        return CGate(control, gate_function(target, *args, **kwargs))
+
+    return cgate
+
+
+def get_sympy_gates():
+    """Map gate name of the abstract format to the equivalent methods of the
+    sympy backend (symbolic simulation).
+    """
+
+    import sympy.physics.quantum.gate as SYMPYGate
+
+    GATE_SYMPY = dict()
+    GATE_SYMPY["H"] = SYMPYGate.HadamardGate
+    GATE_SYMPY["X"] = SYMPYGate.XGate
+    GATE_SYMPY["Y"] = SYMPYGate.YGate
+    GATE_SYMPY["Z"] = SYMPYGate.ZGate
+    GATE_SYMPY["S"] = SYMPYGate.PhaseGate
+    GATE_SYMPY["T"] = SYMPYGate.TGate
+    GATE_SYMPY["PHASE"] = p_gate
+    GATE_SYMPY["SWAP"] = SYMPYGate.SwapGate
+    GATE_SYMPY["RX"] = rx_gate
+    GATE_SYMPY["RY"] = ry_gate
+    GATE_SYMPY["RZ"] = rz_gate
+    GATE_SYMPY["RX"] = rx_gate
+    GATE_SYMPY["CH"] = controlled_gate(SYMPYGate.HadamardGate)
+    GATE_SYMPY["CNOT"] = SYMPYGate.CNotGate
+    GATE_SYMPY["CX"] = SYMPYGate.CNotGate
+    GATE_SYMPY["CY"] = controlled_gate(SYMPYGate.YGate)
+    GATE_SYMPY["CZ"] = controlled_gate(SYMPYGate.ZGate)
+    GATE_SYMPY["CRX"] = controlled_gate(rx_gate)
+    GATE_SYMPY["CRY"] = controlled_gate(ry_gate)
+    GATE_SYMPY["CRZ"] = controlled_gate(rz_gate)
+    GATE_SYMPY["CS"] = controlled_gate(SYMPYGate.PhaseGate)
+    GATE_SYMPY["CT"] = controlled_gate(SYMPYGate.TGate)
+    GATE_SYMPY["CPHASE"] = controlled_gate(p_gate)
+
+    return GATE_SYMPY
+
+
+def translate_c_to_sympy(source_circuit):
+    """Take in an abstract circuit, return a quantum circuit object as defined
+    in the Python SYMPY SDK.
+
+    Args:
+        source_circuit: quantum circuit in the abstract format.
+
+    Returns:
+        SYMPY.circuits.Circuit: quantum circuit in Python SYMPY SDK format.
+    """
+
+    from sympy import symbols
+
+    GATE_SYMPY = get_sympy_gates()
+
+    # Identity as an empty circuit.
+    target_circuit = 1
+
+    # Map the gate information properly.
+    for gate in reversed(source_circuit._gates):
```

Comment:
I expect `in reversed (source_circuit)` to work.

Reply:
With a simple example, I get
```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'Circuit' object is not reversible
```

The `___reversed___` magic method must be defined.

Reply:
Thank you !